src/format.rs

```diff
@@ -532,3 +533,116 @@ pub struct CommandHistory {
+}
+
+/// The conclusion of running `check` or `suggest`
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(tag = "conclusion")]
+pub enum JsonReportConclusion {
+    /// Success! Everything's Good.
+    #[serde(rename = "success")]
+    Success(JsonReportSuccess),
+    /// The violations and audits/exemptions are contradictory!
+    #[serde(rename = "fail (violation)")]
+    FailForViolationConflict(JsonReportFailForViolationConflict),
+    /// The audit failed, here's why and what to do.
+    #[serde(rename = "fail (vetting)")]
+    FailForVet(JsonReportFailForVet),
+}
+
+/// Success! Everything is audited!
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct JsonReportSuccess {
+    /// These packages are fully vetted
+    pub vetted_fully: Vec<JsonPackage>,
+    /// These packages are partially vetted (some audits but relies on an `exemption`).
+    pub vetted_partially: Vec<JsonPackage>,
+    /// These packages are exempted
+    pub vetted_with_exemptions: Vec<JsonPackage>,
+}
+
+/// Failure! The violations and audits/exemptions are contradictory!
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct JsonReportFailForViolationConflict {
+    /// These packages have the following conflicts
+    pub violations: SortedMap<PackageAndVersion, Vec<ViolationConflict>>,
+}
+
+/// Failure! You need more audits!
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct JsonReportFailForVet {
+    /// Here are the problems we found
+    pub failures: Vec<JsonVetFailure>,
+    /// And here are the fixes we recommend
+    pub suggest: Option<JsonSuggest>,
+}
+
+/// Suggested fixes for a FailForVet
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct JsonSuggest {
+    /// Here are the suggestions sorted in the order of priority
+    pub suggestions: Vec<JsonSuggestItem>,
+    /// The same set of suggestions but grouped by the criteria (lists) needed to audit them
+    pub suggest_by_criteria: SortedMap<String, Vec<JsonSuggestItem>>,
+    /// The total number of lines you would need to review to resolve this
+    pub total_lines: u64,
+}
+
+/// This specific package needed the following criteria but doesn't have them!
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct JsonVetFailure {
+    /// The name of the package
+    pub name: PackageName,
+    /// The version of the package
+    pub version: Version,
+    /// The missing criteria
+    pub missing_criteria: Vec<CriteriaName>,
+}
+
+/// We recommend auditing the following package
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct JsonSuggestItem {
+    /// The name of the package
+    pub name: PackageName,
+    /// Any notable parents the package has (can be helpful in giving context to the user)
+    pub notable_parents: String,
+    /// The criteria we recommend auditing the package for
+    pub suggested_criteria: Vec<CriteriaName>,
+    /// The diff (or full version) we recommend auditing
+    pub suggested_diff: DiffRecommendation,
```

Comment:
I realize this is existing, but I don't love how we expose the diff internal representation to the json output. It would be nice if we could have this use a delta-or-version like format like we do for the actual audits.toml format, so that the fact that full diffs are `"null -> 1.0.0"` isn't visible in public APIs.

The raw diffstat being encoded directly in there is also a bit weird.